### ISSUE TITLE
[Feature request] Being able to modify the default `minTime`, `maxTime` and `slotDuration` of fullcalendar with a GUI

### ISSUE THREAD

**Issue by AlexandreBonneau** [nextcloud/calendar#577]
Whenever I upgrade nextcloud, I need to manually fiddle in the `dist/fullcalendar.js` file to modify the *not-so-great* default values.

Using the info from https://fullcalendar.io/docs/agenda/, I usually modify the start (`minTime`) and end (`maxTime`) time of the day (ie. 6:00 and 19:00), and the `slotDuration` to `'00:15:00'` instead of `'00:30:00'`.

Would it be possible to have a simple GUI to do that?

Context: I use a 15 minutes slot duration because I have lots of appointments and those are set every 15 minutes.

**Comment by georgehrke** [comment]
cc @jancborchardt about the defaults.

**Comment by jancborchardt** [comment]
The default of 30min is pretty standard. Introducing a setting for that would obscure more important settings.

Can you provide screenshots of how (if) other widely-used and well-designed software provides an option for that?

(Instead of a visible setting, one thing we could do is notice if there are a lot of <30min appointments, and adjust the default time accordingly.)

**Comment by AlexandreBonneau** [comment]
The 30 minutes duration is indeed standard when you use the calendar to setup meetings and the like.
However, as soon as you use the calendar to set appointments (for instance appointments for a doctor or hospital office), then 15 minutes is mandatory (and I even know ophthalmologists who set appointments every 6 minutes!).

This is important for them, but the two other settings (`minTime` and `maxTime`) are used quite often too: you really do not need to clutter your view with non-office hours. `00:00` to `06:00` and `19:00` to `23:59` are usually never used in most offices, so why bother showing those (or even the scrollbar)?

### About other calendars
Korganizer for instance allows to use `ctrl + mousewheel` to modify the size of the viewport, which allows to actually see the event titles.
This does not change the `slotDuration` per se, since it already default to 15 minutes (see the blue rectangle on the screenshot ; this is the default when clicking on the view), but only zoom in/out.

Before:
![korganizer_ss1](https://user-images.githubusercontent.com/7248877/29947627-5528bc1c-8e46-11e7-93e4-eeb6507a79a7.png)

After:
![korganizer_ss2](https://user-images.githubusercontent.com/7248877/29947626-5519cd1a-8e46-11e7-86a4-92b846408f53.png)

A real-life usecase:
![nextcloud_calendar](https://user-images.githubusercontent.com/7248877/29947820-50a3fd22-8e47-11e7-9472-8316fce3cf07.png)

**Comment by jancborchardt** [comment]
Right, so by default the day or week view should be scrolled to 6 (or 8) anyway. @georgehrke?

Zooming should be possible using Ctrl +/- or mousewheel in the browser.

About slot duration, again we could check if it's modified a lot, and take the shorter (or longer) time as a new default.

**Comment by AlexandreBonneau** [comment]
Well, fullcalendar does have an option `scrollTime` (cf. the [doc](https://fullcalendar.io/docs/agenda/) that determines how far down the scroll pane is initially scrolled down.
But that misses the point : why scroll (and have a scroll bar) when you are sure you'll never set an appointment at 2am?

Zooming in the browser also zoom the sidebars, the header, the incidence editor and everything else ; that's not good enough in my opinion.

I still think that in the settings, there should be a simple form with a few `<input>` element to change the `slotDuration`,`minTime` and `maxTime` settings.

Pretty please ;)

**Comment by jancborchardt** [comment]
We won't introduce settings for everything like that ;) see this great blog post about it: http://ometer.com/preferences.html

As said, the slot duration could be automatically adjusted.

And about scrolling to the bottom:

> why scroll (and have a scroll bar) when you are sure you'll never set an appointment at 2am?

Cause we are not only designing an interface specifically for the usecase you describe. :) Automatically scrolling to 6/8 or so and still leaving the possibility to scroll up is equally helpful while still being versatile.

**Comment by AlexandreBonneau** [comment]
I have the same concern about UI clutter and creating too many preferences options, but the reality is that currently the settings for the calendar are very, very scarce.
Currently the modifiable options are:
1. Use the simple editor
2. Display week numbers

That's it.
*('Import a calendar' and the caldav urls are not options per se (and should not be displayed here in my opinion, but that's another UX problem)).*

So to me the number of preferences here is not the problem.

However, one could imagine that the `slotDuration`, `scrollTo`, `minTime` and `maxTime` are the *core* variables that affect the display of any calendar quite a lot.
To that I would also add the option to choose if the first day of the week is Sunday or Monday, for easy access, and then you would get the very basic configurability needed for a generic calendar.

**Comment by georgehrke** [comment]
> To that I would also add the option to choose if the first day of the week is Sunday or Monday,

right now thats determined based on your language but we are already working on adding a dedicated locale input to the personal settings page

**Comment by jancborchardt** [comment]
@georgehrke is there a current scrollTo value set? If not, can we put it to 6?

**Comment by georgehrke** [comment]
We currently don't set scrollTo, but that's also not necessary because full calendars default is 6
https://fullcalendar.io/docs/agenda/scrollTime/

**Comment by georgehrke** [comment]
To bring this issue back to life:

What about adapting this automatically depending on the user's events? 

Like:
- are there multiple hours where there are at least 4 (consecutive) events?
- do events commonly start in the second half of each half hour section (at xx:20, xx:50)

**Comment by AlexandreBonneau** [comment]
Well, that would need a *lot* more work to analyze the events (and when you have 600+ each week (think ophthalmologist 'basic' 6 min rendez-vous, so 10 rdv per hour, 10 hours a day, 6 days a week); I'm guessing the calculations could get pretty heavy).

Also, if you go with:
> are there multiple hours where there are at least 4 (consecutive) events?

Then you would lose the coherence of a workplace where the hours are always the same, but it's entirely possible that in one particular week, all the events could start later; it would still be logical to set the same 'start hour', even for that particular week.

I still think the 4 to 5 options are not 'cluter', specially since they would be hidden by default and shown only when the user clicks on the `settings` button.

About @jancborchardt comment:
> Automatically scrolling to 6/8 or so and still leaving the possibility to scroll up is equally helpful while still being versatile.

That could only work if there would be a way to directly define the height size of the hours (ie. 1 hours = 50px).

**Comment by jancborchardt** [comment]
As said above multiple times:

> one thing we could do is notice if there are a lot of <30min appointments, and adjust the default time accordingly.

Let’s take small steps, and that’s probably the first one. ;) Simply remembering what was done last.

**Comment by jancborchardt** [comment]
@abate not that I know of, but:
- :man_technologist: You are very welcome to get involved and work on the feature! :) As said a good solution is remembering the last time length that was used, when the default was modified.
- :money_with_wings: If you don’t have time to spend on code, please donate to the [bounty on this issue](https://www.bountysource.com/issues/48623943-feature-request-being-able-to-modify-the-default-mintime-maxtime-and-slotduration-of-fullcalendar-with-a-gui?utm_campaign=plugin&utm_content=tracker%2F45525646&utm_medium=issues&utm_source=github)

:tada:

**Comment by fernandoc2021** [comment]
![screenshot at 2018-11-26 12-46-11](https://user-images.githubusercontent.com/45299139/49032070-ad8f8100-f179-11e8-8294-2cb320baad1e.png)
@AlexandreBonneau you mean something like this? This is under Open-xchange Calendar Settings

**Comment by fernandoc2021** [comment]
Once set (Under Open-xchange) the scrollTime setting is assumed by default based on the Start and End of working time.

**Comment by jancborchardt** [comment]
Sorry, we are not going to introduce settings for every little thing as other software does. This makes everything unnecessarily complicated. Nextcloud is simple and should automatically adjust slot times like that to your recently used ones.

**Comment by abate** [comment]
I agree with the feeling of not create interfaces that are too complicated. I'm participating to this discussion because the CEO of my company refuses to use nextcloud **just** because the calendar does not work exactly as on exchange (?!**$#%). **every little thing** ok, it's a bit too much, but at least the office hours settings ?

**Comment by georgehrke** [comment]
There are plans to implement business hours in order to make them available for Free/Busy planning.
I would be fine with automatically hiding hours outside the business hours as long as there are no events in there.

slotDuration should automatically be set based on the number of events / their starting time / duration. No need for a setting.

**Comment by AlexandreBonneau** [comment]
> I would be fine with automatically hiding hours outside the business hours as long as there are no events in there.
> 
> slotDuration should automatically be set based on the number of events / their starting time / duration. No need for a setting.

Does this means then that if you show only the 6h to 18h interval, then the slotDuration will automatically default to 15 minutes since you would have to display half the number of hours in a given day? If yes, great!

**Comment by jancborchardt** [comment]
> the CEO of my company refuses to use nextcloud just because the calendar does not work exactly as on exchange

@abate does your company already have a [Nextcloud enterprise subscription](https://nextcloud.com/enterprise/)? In that case you could use the support channels offered so we can prioritize this.

**Comment by georgehrke** [comment]
Summary:
- when you have availability set, automatically scroll to first hour of availability.
- When hitting CMD / CTRL + Scrolling, zoom the grid view and save the durationSlot for future use

As a later step:
- automatically determine a good durationSlot based on amount of events

**Comment by doitux** [comment]
Is there a way to manually adjust slotDuration to '00:30:00' in the productive code of 2.0.1 in between?

**Comment by georgehrke** [comment]
@doitux Clone the source, adjust [this line](https://github.com/nextcloud/calendar/blob/master/src/views/Calendar.vue#L178), run `make` and `make appstore`

**Comment by doitux** [comment]
@georgehrke Thanks for the fast reply. 
When i run "make" i get this error:
--------------------------------------------------------
Test Suites: 50 passed, 50 total
Tests:       163 passed, 163 total
Snapshots:   0 total
Time:        14.382s
Ran all test suites.
phpunit -c phpunit.xml
Could not read "phpunit.xml".
make: *** [Makefile:48: test-php] Fehler 1
--------------------------------------------------------

Doing "make appstore" build the tarball without problems. Should i ignore the make error? 
What is the best way to install this tarbal in my productive system? Is there a way to upload and install? Or should i manually unpack and overwrite the apps/Calendar directory?

**Comment by doitux** [comment]
manually unpacking works fine ;-)

**Comment by Malspherus** [comment]
Seems you can also search for `return"00:15:00"` in `apps/calendar/js/calendar.js` and just change it there (if this is not recommended for some reason please let me know). Thank you for the link @georgehrke!

I have a slightly different reason for wanting to change this default, having slotDuration with one "line" for 15min is wasting a lot of screen-space on my laptop and forces me to scroll all the time. Zooming the whole page causes text to get unreadable before I can reach a proper event-size.

**Comment by meaz** [comment]
It would be nice to have it as an option.

**Comment by alexanderdd** [comment]
@tcitworld why was it removed from the milestone?

Slot durations can now be adjusted in settings, that's good, and it wasn't the case when this bug was opened.

But day/week view still starts at 6:00 am, no way to change that, no automatic adjustment. IMO this needs to improve.

**Comment by tcitworld** [comment]
We removed issues that aren't scheduled from milestones, otherwise it's just frustrating to see the issue being postponed for ever like above.

**Comment by ChristophWurst** [comment]
@alexanderdd fyi see https://github.com/nextcloud/groupware/blob/main/processes/milestones.md. I tried to formalize this a bit. As @tcitworld  said as long as this is not planned we'd just move from one milestone to the next. Hence we unassigned the many tickets that had a milestone.
